Code/Framework/AzCore/AzCore/DOM/DomPrefixTree.h

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+#include <AzCore/DOM/DomPath.h>
+#include <AzCore/std/containers/map.h>
+#include <AzCore/std/containers/stack.h>
+#include <AzCore/std/optional.h>
+
+namespace AZ::Dom
+{
+    //! Specifies how a patch matches against a DomPrefixTree
+    enum class PrefixTreeMatch
+    {
+        //! Only an exact path will match.
+        //! For the path "/foo/bar" only "/foo/bar" will match while
+        //! "/foo" and "/foo/bar/baz" will not.
+        ExactPath,
+        //! The path, and any of its subpaths, will match.
+        //! For the path "/foo/bar" both "/foo/bar" and any subpaths like
+        //! "/foo/bar/0" will match, while "/foo" and orthogonal paths like
+        //! "/bar" will not
+        PathAndSubpaths,
+        //! Any of the path's subpaths will match, excepting the path itself.
+        //! For the path "/foo/bar", "/foo/bar" will not match but "/foo/bar/0"
+        //! will.
+        SubpathsOnly,
+    };
+
+    //! A prefix tree that maps DOM paths to some arbitrary value.
+    template<class T>
+    class DomPrefixTree
+    {
+    public:
+        DomPrefixTree() = default;
+        DomPrefixTree(const DomPrefixTree&) = default;
+        DomPrefixTree(DomPrefixTree&&) = default;
+        explicit DomPrefixTree(AZStd::initializer_list<AZStd::pair<Path, T>> init);
+
+        //! Visits a path and calls a visitor for each matching path and value.
+        void VisitPath(const Path& path, PrefixTreeMatch match, AZStd::function<void(const Path&, const T&)> visitor) const;
```

Comment:
Since the function won't be stored locally it's better to take it by const reference.